test/threadsafe_function_ex/test/basic.cc

```diff
@@ -0,0 +1,433 @@
+                      1,               // size_t initialThreadCount,
+                      nullptr,         // ContextType* context
+                      base::Finalizer, // Finalizer finalizer
+                      &_deferred       // FinalizerDataType* data
+    );
+  }
+
+  static std::array<ClassPropertyDescriptor<TSFNWrap>, 2> InstanceMethods() {
+    return {{InstanceMethod("release", &TSFNWrap::Release),
+             InstanceMethod("call", &TSFNWrap::Call)}};
+  }
+
+  Napi::Value Call(const CallbackInfo &info) {
+    auto data = new DataType{Promise::Deferred::New(info.Env())};
+    _tsfn.NonBlockingCall(data);
+    return data->deferred.Promise();
+  };
+};
+
+#endif
+} // namespace empty
+
+namespace existing {
+
+// Data passed (as pointer) to [Non]BlockingCall
+struct DataType {
+  Promise::Deferred deferred;
+  bool reject;
+};
+
+// CallJs callback function provided to `napi_create_threadsafe_function`. It is
+// _NOT_ used by `Napi::ThreadSafeFunctionEx<>`, which is why these arguments
+// are napi_*.
+static void CallJs(napi_env env, napi_value jsCallback, void * /*context*/,
+                   void *data) {
+  DataType *casted = static_cast<DataType *>(data);
+  if (env != nullptr) {
+    if (jsCallback != nullptr) {
+      Function(env, jsCallback).Call(0, nullptr);
+    }
+    if (data != nullptr) {
+      if (casted->reject) {
+        casted->deferred.Reject(
+            String::New(env, "The CallJs has rejected the promise"));
+      } else {
+        casted->deferred.Resolve(
+            String::New(env, "The CallJs has resolved the promise"));
+      }
+    }
+  }
+  if (casted != nullptr) {
+    delete casted;
+  }
+}
+
+// This test creates a native napi_threadsafe_function itself, whose `context`
+// parameter is the `TSFNWrap` object. We forward-declare, so we can use
+// it as an argument inside `ThreadSafeFunctionEx<>`. This also allows us to
+// statically get the correct type when using `tsfn.GetContext()`. The converse
+// is true: if the ContextType does _not_ match that provided to the underlying
+// napi_create_threadsafe_function, then the static type will be incorrect.
+class TSFNWrap;
+
+// Context of the TSFN.
+using ContextType = TSFNWrap;
+
+// Full type of our ThreadSafeFunctionEx
+using TSFN = ThreadSafeFunctionEx<ContextType, DataType>;
+using base = tsfnutil::TSFNWrapBase<TSFNWrap, ContextType, TSFN>;
+
+// A JS-accessible wrap that holds a TSFN.
+class TSFNWrap : public base {
+public:
+  TSFNWrap(const CallbackInfo &info) : base(info) {
+
+    auto env = info.Env();
+
+    if (info.Length() < 1 || !info[0].IsFunction()) {
+      NAPI_THROW_VOID(Napi::TypeError::New(
+          env, "Invalid arguments: Expected arg0 = function"));
+    }
+
+    napi_threadsafe_function napi_tsfn;
+
+    // A threadsafe function on N-API 4 still requires a callback function, so
+    // this uses the `EmptyFunctionFactory` helper method to return a no-op
+    // Function on N-API 5+.
```

Comment:
Is this comment correct ?  I thought the no-op function was only needed for N-API 4 and below?

Reply:
This is no longer relevant as this file has been deleted.